src/wp-includes/class-wp-theme-json-resolver.php

```diff
@@ -531,14 +559,18 @@ public static function get_user_data() {
 				isset( $decoded_data['isGlobalStylesUserThemeJSON'] ) &&
 				$decoded_data['isGlobalStylesUserThemeJSON']
 			) {
-				unset( $decoded_data['isGlobalStylesUserThemeJSON'] );
 				$config = $decoded_data;
 			}
 		}
 
 		/** This filter is documented in wp-includes/class-wp-theme-json-resolver.php */
-		$theme_json   = apply_filters( 'wp_theme_json_data_user', new WP_Theme_JSON_Data( $config, 'custom' ) );
-		static::$user = $theme_json->get_theme_json();
+		$theme_json = apply_filters( 'wp_theme_json_data_user', new WP_Theme_JSON_Data( $config, 'custom' ) );
+		$config     = $theme_json->get_data();
+
+		// Needs to be set for schema migrations of user data.
+		$config['isGlobalStylesUserThemeJSON'] = true;
```

Comment:
This is one of the things I haven't been able to test, and want to leave a note to clarifying it. @ajlende can you provide reproduction steps for this? How does it fail without this?

Reply:
Note this also changed recently at https://github.com/WordPress/wordpress-develop/pull/6271 to remove the extra-parsing.

Reply:
It's used during the migration to prevent adding `"defaultFontSizes": false` and `"defaultSpacingSizes: false` to empty user data since those values aren't configurable by the user in the global styles UI.

See https://github.com/WordPress/wordpress-develop/pull/6616/files#diff-7bab66e2c26e312df49f13ff601b4ab55601e876c21abdcf21cd884def623226R112-R121.

Reply:
I also responded on the Gutenberg PR for https://github.com/WordPress/wordpress-develop/pull/6271. See https://github.com/WordPress/gutenberg/pull/61262#discussion_r1610736278.

Reply:
Approved the follow-up you prepared https://github.com/WordPress/gutenberg/pull/62305 It's a nice one, thank you!

Reply:
Backport for that follow-up at https://github.com/WordPress/wordpress-develop/pull/6737